File: 第11章 竞赛趣题/11.15 颠倒车牌号/颠倒车牌号.py
'''
程序：颠倒车牌号
作者：苏秦@小海豚科学馆公众号
来源：图书《趣味编程三剑客：从Scratch迈向Python和C++》
'''
def main():
    '''颠倒车牌号'''
    total = 0
    for i in range(0, 100000):
        # 用零补足5位数
        number = '0' * (5 - len(str(i))) + str(i)
        # 颠倒数字
        reverse = ''
        for digit in number:
            reverse = '0100009086'[int(digit)] + reverse
        # 找到符合的数字
        if number == reverse:
            print(number)
            total += 1
    # 输出符合的车牌号数量
    print(total)

if __name__ == '__main__':
    main()

File: 第11章 竞赛趣题/11.15 颠倒车牌号/颠倒车牌号.cpp
/*
程序：颠倒车牌号 
作者：苏秦@小海豚科学馆公众号
来源：图书《趣味编程三剑客：从Scratch迈向Python和C++》
*/
#include <bits/stdc++.h>
using namespace std;

// 用零补足5位 
string fill_zero(int n)
{
	stringstream ss; ss << n;
	string num_str; ss >> num_str;
	
	if (n > 9999)
		return num_str;
	
	string zeros = "";
	for (int i = 0; i < 5 - num_str.size(); i++)
		zeros += "0";
	num_str = zeros + num_str;
	
	return num_str;
}

// 颠倒数字 
string reverse_num(string num_str)
{
	string digits = "0100009086";
	string reverse = "";
	for (int i = 0; i < num_str.size(); i++)
		reverse = digits[num_str[i] - '0'] + reverse;
	return reverse;
}

// 颠倒车牌号
int main()
{
	int total = 0;
	for (int i = 0; i < 100000; i++) {
		string n = fill_zero(i);
		string reverse = reverse_num(n);
		if (n == reverse) {
			cout << n << endl;
			total += 1;
		}
	}
	cout << total << endl;
	return 0;
}
